src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -966,21 +974,39 @@ private static bool IsValidCloseStatus(WebSocketCloseStatus closeStatus)
                 throw new WebSocketException(error, innerException);
             }
 
-            /// <summary>Reads a message header from the network.</summary>
-            /// <param name="cancellationToken">The CancellationToken used to cancel the websocket.</param>
-            /// <returns>The read header, or null if we couldn't read one.</returns>
-            private async Task<MessageHeader?> ReadMessageHeaderAsync(CancellationToken cancellationToken)
+            private async Task<bool> EnsureBufferContainsHeaderAsync(CancellationToken cancellationToken)
             {
-                // Read the first two bytes of the header, which gives us the opcode, FIN, reserved bits, and mask state
-                if (_receiveBufferCount < 2)
+                if (_receiveBufferCount < MaxMessageHeaderLength)
                 {
-                    await EnsureBufferContainsAsync(2, cancellationToken, throwOnPrematureClosure: false).ConfigureAwait(false);
+                    // Make sure we have the first two bytes, which includes the start of the payload length
                     if (_receiveBufferCount < 2)
                     {
-                       return null;
+                        await EnsureBufferContainsAsync(2, cancellationToken, throwOnPrematureClosure: false).ConfigureAwait(false);
+                        if (_receiveBufferCount < 2)
+                        {
+                            return false;
+                        }
+                    }
+
+                    // Make sure we have the full header based on the payload length
+                    long payloadLength = _receiveBuffer[_receiveBufferOffset + 1] & 0x7F;
+                    if (payloadLength > 125)
+                    {
+                        await EnsureBufferContainsAsync(payloadLength == 126 ? 4 : 10, cancellationToken).ConfigureAwait(false);
```

Comment:
This rings a bell, the variable sized length stuff... it might be useful to bring part of that comment here.  And maybe replace the 4 and 10 with math using sizeof().